File: etherscan_common.py
"""
Shared validation and utility functions for Etherscan token balance queries.
This module provides common functionality used across multiple implementations.
"""

import re
import json
import os
from urllib.parse import urlencode

# Ethereum address validation pattern
ETHEREUM_ADDRESS_PATTERN = re.compile(r'^0x[a-fA-F0-9]{40}$')

# Messages cache
_MESSAGES = None


def load_messages():
    """
    Load shared messages from JSON file.
    
    Returns:
        dict: Messages data
    """
    global _MESSAGES
    if _MESSAGES is None:
        messages_path = os.path.join(os.path.dirname(__file__), 'etherscan-messages.json')
        try:
            with open(messages_path, 'r') as f:
                _MESSAGES = json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            # Fallback to empty dict if messages file not found
            _MESSAGES = {"errors": {}, "status": {}, "labels": {}}
    return _MESSAGES


def validate_ethereum_address(address):
    """
    Validate Ethereum address format.
    
    Args:
        address (str): Ethereum address to validate
        
    Returns:
        bool: True if valid, False otherwise
    """
    return bool(ETHEREUM_ADDRESS_PATTERN.match(address))


def load_config(config_path=None):
    """
    Load shared configuration from JSON file.
    
    Args:
        config_path (str, optional): Path to config file. If None, uses default location.
        
    Returns:
        dict: Configuration data
        
    Raises:
        FileNotFoundError: If config file doesn't exist
        json.JSONDecodeError: If config file is not valid JSON
    """
    if config_path is None:
        config_path = os.path.join(os.path.dirname(__file__), 'etherscan-api-config.json')
    
    try:
        with open(config_path, 'r') as f:
            return json.load(f)
    except FileNotFoundError:
        raise FileNotFoundError(
            f"Configuration file not found: {config_path}\n"
            "Please ensure etherscan-api-config.json exists in the script directory."
        )
    except json.JSONDecodeError as e:
        raise json.JSONDecodeError(
            f"Invalid JSON in configuration file: {config_path}",
            e.doc, e.pos
        )


def build_api_params(config, address, api_key, chain_id=None, page=None, offset=None):
    """
    Build API request parameters from config and user inputs.
    
    Args:
        config (dict): Shared configuration
        address (str): Ethereum address to query
        api_key (str): Etherscan API key
        chain_id (int, optional): Chain ID
        page (int, optional): Page number
        offset (int, optional): Results per page
        
    Returns:
        dict: API request parameters
    """
    params = {
        "chainid": chain_id or config['defaultChainId'],
        "module": config['module'],
        "action": config['action'],
        "address": address,
        "page": page or config['defaultPage'],
        "offset": offset or config['defaultOffset'],
        "apikey": api_key
    }
    return params


def build_api_url(config, params):
    """
    Build full API URL with parameters.
    
    Args:
        config (dict): Shared configuration
        params (dict): API request parameters
        
    Returns:
        str: Full API URL with query parameters
    """
    base_url = config['apiBaseUrl']
    query_string = urlencode(params)
    return f"{base_url}?{query_string}"


def format_token_balance(token_data):
    """
    Format token balance data for display.
    
    Args:
        token_data (dict): Token data from API response
        
    Returns:
        str: Formatted token information
    """
    lines = []
    lines.append(f"  Token: {token_data.get('TokenName', 'Unknown')}")
    lines.append(f"  Symbol: {token_data.get('TokenSymbol', 'N/A')}")
    lines.append(f"  Address: {token_data.get('TokenAddress', 'N/A')}")
    lines.append(f"  Quantity: {token_data.get('TokenQuantity', '0')}")
    lines.append(f"  Divisor: {token_data.get('TokenDivisor', '18')}")
    return "\n".join(lines)


def is_response_successful(response):
    """
    Check if API response indicates success.
    
    Args:
        response (dict): API response data
        
    Returns:
        bool: True if successful, False otherwise
    """
    return response and response.get("status") == "1"


def format_response(response, pretty=False):
    """
    Format API response output.
    
    Args:
        response (dict): API response data
        pretty (bool): Whether to pretty-print the JSON
        
    Returns:
        str: Formatted JSON string
    """
    indent = 2 if pretty else None
    return json.dumps(response, indent=indent)

File: etherscan-common.js
/**
 * Shared validation and utility functions for Etherscan token balance queries.
 * This module provides common functionality used across multiple implementations.
 */

const fs = require('fs');
const path = require('path');

// Cache for config
let _config = null;

/**
 * Get cached configuration.
 * @returns {object} Configuration data
 */
function getConfig() {
  if (!_config) {
    _config = loadConfig();
  }
  return _config;
}

// Load shared messages
let MESSAGES = null;
function loadMessages() {
  if (!MESSAGES) {
    const messagesPath = path.join(__dirname, 'etherscan-messages.json');
    try {
      const messagesData = fs.readFileSync(messagesPath, 'utf8');
      MESSAGES = JSON.parse(messagesData);
    } catch (err) {
      // Fallback to empty object if messages file not found
      MESSAGES = { errors: {}, status: {}, labels: {} };
    }
  }
  return MESSAGES;
}

/**
 * Validate Ethereum address format.
 * @param {string} address - Ethereum address to validate
 * @returns {boolean} True if valid, false otherwise
 */
function validateEthereumAddress(address) {
  const config = getConfig();
  const pattern = new RegExp(config.validationPatterns.ethereumAddress);
  return pattern.test(address);
}

/**
 * Get Ethereum address validation pattern (for backward compatibility).
 * @returns {RegExp} Ethereum address validation pattern
 */
function getEthereumAddressPattern() {
  const config = getConfig();
  return new RegExp(config.validationPatterns.ethereumAddress);
}

// Lazy-loaded pattern for backward compatibility
let _ethereumAddressPattern = null;
Object.defineProperty(exports, 'ETHEREUM_ADDRESS_PATTERN', {
  get: function() {
    if (!_ethereumAddressPattern) {
      _ethereumAddressPattern = getEthereumAddressPattern();
    }
    return _ethereumAddressPattern;
  }
});

/**
 * Load shared configuration from JSON file.
 * @param {string} [configPath] - Path to config file. If null, uses default location.
 * @returns {object} Configuration data
 * @throws {Error} If config file doesn't exist or contains invalid JSON
 */
function loadConfig(configPath = null) {
  if (!configPath) {
    // Try to load etherscan-api-config.json first, fall back to example
    const configFile = path.join(__dirname, 'etherscan-api-config.json');
    const exampleFile = path.join(__dirname, 'etherscan-api-config.example.json');
    
    if (fs.existsSync(configFile)) {
      configPath = configFile;
    } else {
      configPath = exampleFile;
    }
  }
  
  try {
    const configData = fs.readFileSync(configPath, 'utf8');
    const config = JSON.parse(configData);
    // Update cached config
    _config = config;
    return config;
  } catch (err) {
    if (err.code === 'ENOENT') {
      throw new Error(
        `Configuration file not found: ${configPath}\n` +
        'Please ensure etherscan-api-config.json or etherscan-api-config.example.json exists in the script directory.'
      );
    } else if (err instanceof SyntaxError) {
      throw new Error(`Invalid JSON in configuration file: ${configPath}\n${err.message}`);
    }
    throw err;
  }
}

/**
 * Build API request parameters from config and user inputs.
 * @param {object} config - Shared configuration
 * @param {string} address - Ethereum address to query
 * @param {string} apiKey - Etherscan API key
 * @param {number} [chainId] - Chain ID
 * @param {number} [page] - Page number
 * @param {number} [offset] - Results per page
 * @returns {object} API request parameters
 */
function buildApiParams(config, address, apiKey, chainId = null, page = null, offset = null) {
  return {
    chainid: chainId || config.defaultChainId,
    module: config.module,
    action: config.action,
    address: address,
    page: page || config.defaultPage,
    offset: offset || config.defaultOffset,
    apikey: apiKey
  };
}

/**
 * Build full API URL with parameters.
 * @param {object} config - Shared configuration
 * @param {object} params - API request parameters
 * @returns {string} Full API URL
 */
function buildApiUrl(config, params) {
  const url = new URL(config.apiBaseUrl);
  Object.entries(params).forEach(([key, value]) => {
    url.searchParams.append(key, value);
  });
  return url.toString();
}

/**
 * Format token balance data for display.
 * @param {object} tokenData - Token data from API response
 * @returns {string} Formatted token information
 */
function formatTokenBalance(tokenData) {
  const lines = [];
  lines.push(`  Token: ${tokenData.TokenName || 'Unknown'}`);
  lines.push(`  Symbol: ${tokenData.TokenSymbol || 'N/A'}`);
  lines.push(`  Address: ${tokenData.TokenAddress || 'N/A'}`);
  lines.push(`  Quantity: ${tokenData.TokenQuantity || '0'}`);
  lines.push(`  Divisor: ${tokenData.TokenDivisor || '18'}`);
  return lines.join('\n');
}

/**
 * Check if API response indicates success.
 * @param {object} response - API response data
 * @returns {boolean} True if successful, false otherwise
 */
function isResponseSuccessful(response) {
  return response && response.status === "1";
}

/**
 * Format API response output based on options.
 * @param {object} response - API response data
 * @param {object} options - Formatting options (e.g., {pretty: true})
 * @returns {string} Formatted output string
 */
function formatResponse(response, options = {}) {
  const indent = options.pretty ? 2 : undefined;
  return JSON.stringify(response, null, indent);
}

module.exports = {
  validateEthereumAddress,
  loadConfig,
  loadMessages,
  buildApiParams,
  buildApiUrl,
  formatTokenBalance,
  isResponseSuccessful,
  formatResponse,
  ETHEREUM_ADDRESS_PATTERN
};
